### PR TITLE
Quiz list error view

### PR DESCRIPTION
**Pull request by IvanKulundzic**
Added an error view for when there's an error fetching quizzes. Improved error handling. Added get quiz button for retrying to fetch quizzes.

image

### CODE REVIEW

File: QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift [hunk 1]
```diff
@@ -17,14 +17,28 @@ final class QuizNetworkDataSource: QuizNetworkDataSourceProtocol {
...
-        try await quizNetworkClient.fetchQuizes()
-            .map { QuizDataModel(from: $0) }
+        var models: [QuizDataModel] = []
+        do {
+            models = try await quizNetworkClient.fetchQuizes()
+                .map { QuizDataModel(from: $0) }
+        } catch {
+            throw RequestError.serverError
+        }
+        return models
```
Comment (dorastipkovic) [inline QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift]: `quizNetworkClient.fetchQuizes()` is already throwing. Unless you want to change the error or do something with it, all you need to do is propagate it

```
try await quizNetworkClient.fetchQuizes()
            .map { QuizDataModel(from: $0) }
```

☝️ This is exactly that. You don't need to throw explicitly.

File: QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift [hunk 2]
```diff
@@ -17,14 +17,28 @@ final class QuizNetworkDataSource: QuizNetworkDataSourceProtocol {
...
-            .map { QuizDataModel(from: $0) }
+
+        do {
+            models = try await quizNetworkClient.fetchQuizes(for: categoryNetworkModel)
+                .map { QuizDataModel(from: $0) }
+        } catch {
+            throw error.self
+        }
+        
+        return models
```
Comment (dorastipkovic) [inline QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift]: Same as above.

File: QuizApp/App/Networking/NetworkClient.swift [hunk 3]
```diff
@@ -15,22 +15,19 @@ final class NetworkClient: NetworkClientProtocol {
...
-
-        guard (200...299).contains(response.statusCode) else {
-            switch response.statusCode {
-            case 400:
-                throw RequestError.badRequest
+        if
+            let httpResponse = response as? HTTPURLResponse,
+            (300...503).contains(httpResponse.statusCode)
+        {
+            switch httpResponse.statusCode {
```
Comment (dorastipkovic) [inline QuizApp/App/Networking/NetworkClient.swift]: Can you explain why you changed this?

File: QuizApp/App/Networking/Quiz/QuizNetworkClient.swift [hunk 4]
```diff
@@ -29,7 +29,14 @@ final class QuizNetworkClient: QuizNetworkClientProtocol {
...
 
-        return try await networkClient.executeUrlRequest(request)
+        var models: [QuizNetworkModel] = []
+
+        do {
+           models = try await networkClient.executeUrlRequest(request)
+        } catch {
+            throw RequestError.serverError
+        }
+        return models
```
Comment (dorastipkovic) [inline QuizApp/App/Networking/Quiz/QuizNetworkClient.swift]: Same as above. 
By catching this, you are rewriting all errors thrown in NetworkClient to RequestError.serverError

File: QuizApp/App/Screens/Quiz/QuizListViewController.swift [hunk 5]
```diff
@@ -40,6 +48,9 @@ extension QuizListViewController: UICollectionViewDataSource {
 
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         if collectionView == categoryCollectionView {
+//            let numberOfCategories = quizListViewModel.categories.count
+//            let count = quizListViewModel.quizes.isEmpty ? 0 : numberOfCategories
+//            return count
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewController.swift]: remove

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 6]
```diff
@@ -17,32 +18,43 @@ final class QuizListViewModel {
...
-                self.quizes = quizes
-                    .map { QuizViewModel(from: $0) }
+                    self.quizes = quizes
+                        .map { QuizViewModel(from: $0) }
+                }
+            } catch {
+                hideEmptyStateView = false
             }
         }
     }
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift]: If you mark the method as @MainActor, then dispatching is unneccessary.

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 7]
```diff
@@ -17,32 +18,43 @@ final class QuizListViewModel {
...
+                    hideEmptyStateView = false
+                }
 
-                self.quizes = quizes
-                    .map { QuizViewModel(from: $0) }
             }
-        }
     }
 
+    @MainActor
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift]: Why is this @MainActor?

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 8]
```diff
@@ -17,32 +18,43 @@ final class QuizListViewModel {
 
 extension QuizListViewModel {
 
+    @MainActor
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift:21]: You know have two methods that are basically copy-paste. 
You can have just one private method that will have that code and two public methods that will call that private method with the appropriate arguments.

File: QuizApp/App/Use cases/Quiz/QuizUseCase.swift [hunk 9]
```diff
@@ -17,14 +17,30 @@ final class QuizUseCase: QuizUseCaseProtocol {
...
+        var models: [QuizModel] = []
+        
+        do {
+            models = try await quizNetworkDataSource.fetchQuizes()
+                .map { QuizModel(from: $0) }
+        } catch {
+            throw error.self
+        }
+
+        return models
```
Comment (dorastipkovic) [inline QuizApp/App/Use cases/Quiz/QuizUseCase.swift]: throwing is redundant here

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift (hunk 8 above)
Comment (IvanKulundzic) [reply QuizApp/App/Screens/Quiz/QuizListViewModel.swift:21]: I will fix that in the sections PR, so I don't have to solve merge conflicts later. :)

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift (hunk 7 above)
Comment (IvanKulundzic) [reply QuizApp/App/Screens/Quiz/QuizListViewModel.swift]: As I've marked `fetchQuiz(for category: CategoryModel)` with` @MainActor` and that method calls `fetchInitialQuiz()`, compiler is not happy with an error `Call to main actor-isolated instance method 'fetchQuiz(for:)' in a synchronous nonisolated context`. 

I will refactor those methods in the sections PR, so will look into this more.

File: QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift [hunk 10]
```diff
@@ -17,14 +17,22 @@ final class QuizNetworkDataSource: QuizNetworkDataSourceProtocol {
     }
 
     func fetchQuizes() async throws -> [QuizDataModel] {
-        try await quizNetworkClient.fetchQuizes()
+        var models: [QuizDataModel] = []
+
+        models = try await quizNetworkClient.fetchQuizes()
             .map { QuizDataModel(from: $0) }
+
+        return models
```
Comment (dorastipkovic) [inline QuizApp/App/Data sources/Quiz/QuizNetworkDataSource.swift]:
```suggestion
        return try await quizNetworkClient.fetchQuizes()
            .map { QuizDataModel(from: $0) }
```

Apply everywhere

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/App/Networking/NetworkClient.swift (hunk 3 above)
Comment (IvanKulundzic) [reply QuizApp/App/Networking/NetworkClient.swift]: I was checking something and forgot about it.

Comment (dorastipkovic) [reply QuizApp/App/Networking/NetworkClient.swift]: Then revert to the old way with guard.

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 11]
```diff
@@ -17,32 +18,40 @@ final class QuizListViewModel {
...
-            DispatchQueue.main.async { [weak self] in
-                guard let self = self else { return }
+            do {
+                let quizes = try await quizUseCase.fetchQuizes(for: category)
 
                 self.quizes = quizes
                     .map { QuizViewModel(from: $0) }
+
+            } catch {
+                hideEmptyStateView = false
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift:32]: Do you need to also set `hideEmptyStateView` to true when you successfully fetch quizzes?

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 12]
```diff
@@ -17,32 +18,40 @@ final class QuizListViewModel {
...
             }
         }
     }
 
-    func fetchAllQuizes() {
+    func fetchAllQuizes() throws {
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift]: Do you need `throws` here?

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/App/Screens/Quiz/QuizListViewController.swift [hunk 13]
```diff
@@ -3,11 +3,19 @@ import Combine
 
 final class QuizListViewController: UIViewController {
 
-    private var cancellables = Set<AnyCancellable>()
+
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewController.swift]: extra newline

File: QuizApp/App/Screens/Quiz/QuizListViewModel.swift [hunk 14]
```diff
@@ -18,34 +19,52 @@ final class QuizListViewModel {
 
 extension QuizListViewModel {
 
+
```
Comment (dorastipkovic) [inline QuizApp/App/Screens/Quiz/QuizListViewModel.swift]: extra newline

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

**Review by dorastipkovic — APPROVED** [review]

**Pull request merged**